skimage/feature/fisher_vector.py

```diff
@@ -0,0 +1,261 @@
+               descriptors for a single image in your dataset, or
+               perhaps a test image).
+
+Reference: Perronnin, F. and Dance, C. Fisher kernels on Visual Vocabularies
+           for Image Categorization, IEEE Conference on Computer Vision and
+           Pattern Recognition, 2007
+
+Origin Author: Dan Oneata (Author of the original implementation for the Fisher
+vector computation using scikit-learn and NumPy. Subsequently ported to
+scikit-image (here) by other authors.)
+"""
+import numpy as np
+
+
+class FisherVectorException(Exception):
+    pass
+
+
+class DescriptorException(FisherVectorException):
+    pass
+
+
+def learn_gmm(descriptors, *, n_modes=32, gm_args=None):
+    """Function to estimate a GMM given a set of descriptors and number of
+    modes (i.e. Gaussians). This function is essentially a wrapper around
+    the scikit-learn implementation of GMM, namely the
+    :func:`sklearn.mixture.GaussianMixture` class.
+
+    Due to the nature of the Fisher vector, the only enforced parameter of the
+    underlying scikit-learn class is the covariance_type, which must be 'diag'.
+
+    There is no simple way to know what value to use for `n_modes` a-priori.
+    Typically, the value is usually one of ``{16, 32, 64, 128}``. One may train
+    a few GMMs and choose the one that maximises the log probability of the
+    GMM, or choose `n_modes` such that the downstream classifier trained on
+    the resultant Fisher vectors has maximal performance.
+
+    Parameters
+    -----------
+    descriptors : np.ndarray (N, M) or list [(N1, M), (N2, M), ...]
+        List of NumPy arrays, or a single NumPy array, of the descriptors
+        used to estimate the GMM. The reason a list of NumPy arrays is
+        permissible is because often when using a Fisher vector encoding,
+        descriptors/vectors are computed separately for each sample/image in
+        the dataset, such as SIFT vectors for each image. If a list if passed
+        in, then each element must be a NumPy array in which the number of
+        rows may differ (e.g. different number of SIFT vector for each image),
+        but the number of columns for each must be the same (i.e. the
+        dimensionality must be the same).
+    n_modes : int
+        The number of modes/Gaussians to estimate during the GMM estimate.
+    gm_args : dict
+        Keyword arguments that can be passed into the underlying scikit-learn
+        :func:`sklearn.mixture.GaussianMixture` class.
+
+    Returns
+    -------
+    gmm : :func:`sklearn.mixture.GaussianMixture`
+        The estimated GMM object, which contains the necessary parameters
+        needed to compute the Fisher vector.
+
+    References
+    -----------
+    .. [1] https://scikit-learn.org/stable/modules/generated/sklearn.mixture.GaussianMixture.html
+
+    Examples
+    --------
+    >>> from skimage.feature import fisher_vector
+    >>> sift_for_images = [np.random.random((10, 128)) for _ in range(10)]
```

Comment:
Let's check if this is in line with #5357 :eyes:

Reply:
Done